Perform first healthcheck on startup without waiting the specified interval

yongtang opened the pull request:
**- What I did**
This fix tries to address the issue raised in #33410 where the first healthcheck was only performed after the specified interval.

That might be problematic in case the interval is long yet, as is specified in #33410.

**- How I did it**

This fix make the following changes:
1. The first healthcheck is performed on startup (at time 0)
2. The first healthcheck is only valid when start-period < 0

The changes allows the backward-compatibility while at the same
time it is possbile to address the issue raised in #33410.

**- How to verify it**

Additional test cases have been added.

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

![cute-kitten-kittens-22438020-480-360](https://user-images.githubusercontent.com/6932348/27765938-e1de92b4-5e74-11e7-8d12-5a21f4add2ea.jpg)


This fix fixes #33410.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

aaronlehmann commented on integration-cli/docker_cli_health_test.go:
```diff
@@ -138,6 +138,26 @@ func (s *DockerSuite) TestHealth(c *check.C) {
+		RUN /bin/sleep 1
+		RUN echo OK > /status
+		STOPSIGNAL SIGKILL
+                HEALTHCHECK --interval=1s --timeout=30s --start-period=-1s \
```
Looks like the indentation is not consistent here.

yongtang replied:
@aaronlehmann Thanks. The PR has been updated.

aaronlehmann commented on integration-cli/docker_cli_health_test.go:
```diff
@@ -138,6 +138,26 @@ func (s *DockerSuite) TestHealth(c *check.C) {
+		STOPSIGNAL SIGKILL
+                HEALTHCHECK --interval=1s --timeout=30s --start-period=-1s \
+                  CMD cat /status`))
+	_, _ = dockerCmd(c, "run", "-d", "--name=healthcheck_start",
```
No need for `_, _ = `

yongtang replied:
Thanks. Done.

aaronlehmann commented on daemon/health.go:
```diff
@@ -148,12 +148,20 @@ func handleProbeResult(d *Daemon, c *container.Container, result *types.Healthch
+			if result.Start.IsZero() {
+				// In case this is the first healthcheck, the failing streak
+				// will only be incremented when startPeriod < 0
+				if c.Config.Healthcheck.StartPeriod >= 0 {
```
It's a bit weird that you have to set `StartPeriod` to a negative value to get the health check to run at startup. I would expect it to do this when `StartPeriod` is 0. I know there are backward compatibility issues, though.

aaronlehmann replied:
Perhaps it would be clearer to introduce a new boolean flag for this, instead of overloading `StartPeriod` with a magic value.

yongtang replied:
Thanks @aaronlehmann. The PR has been updated with a flag `--nowait` (True/False) to cover the scenario. Please take a look.

thaJeztah commented:
ping @yongtang

yongtang committed:
Perform first healthcheck on startup without waiting the specified interval

This fix tries to address the issue raised in #33410 where the first
healthcheck was only performed after the specified interval.

That might be problematic in case the interval is long yet, as is
specified in #33410.

This fix make the following changes:
1. The first healthcheck is performed on startup (at time 0)
2. The first healthcheck is only valid when no-wait = true

The changes allows the backward-compatibility while at the same
time it is possbile to address the issue raised in #33410.

Additional test cases have been added.

This fix fixes #33410.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@aaronlehmann The PR has been updated. Please take a look and let me know if there are any issues.

aaronlehmann commented:
Thanks, I like this design a lot more. The naming seems fine to me, but others can feel free to comment if they think of a flag name they prefer to `--no-wait`.

Design LGTM

aaronlehmann commented on builder/dockerfile/dispatchers.go:
```diff
@@ -591,6 +591,7 @@ func healthcheck(req dispatchRequest) error {
 		flTimeout := req.flags.AddString("timeout", "")
 		flStartPeriod := req.flags.AddString("start-period", "")
 		flRetries := req.flags.AddString("retries", "")
+		flNoWait := req.flags.AddBool("nowait", false)
```
Didn't realize this wasn't hyphenated. Wouldn't `no-wait` be more consistent with the other flags?

yongtang commented:
Thanks @aaronlehmann for the review. The PR has been updated with `no-wait`.

cpuguy83 commented:
I wonder if we can just have a whole separate configuration for the startup protocol.
For instance, even if the healtcheck is performed immediately on startup, what if it fails because something wasn't quite ready, and then it needs to wait the entire interval period anyway.

thaJeztah commented:
@yongtang can you have a look at @cpuguy83's comment?

cpuguy83 commented:
@rdxmb The problem is I don't believe this is a good solution. It may be a solution that works for some case that you currently have but that does not make it a good one.

dtakken commented:
I am with @cpuguy83 on this one. I also have a bunch of containers that need a second or so before they are fully running. They should be able to go into healthy state within seconds, but there is no way to make that happen without setting a short check interval. This pull request still would not fix that.

Ideal solution in my case would be if a health check would happen right after the start-period. Then, I can choose a a start period of, say, 5 seconds and a check interval of 5 minutes and still have the container go healthy in seconds.

Some sort of a --wait-for-start option that triggers this behavior would be backward compatible and cover all discussed use cases, right?

dtakken commented:
I am with @cpuguy83 on this one. I also have a couple of containers that need a second to start. These should be able to get into healthy state within seconds without using a really short check interval. Currently, there is no way to do this, and this pull request will not fix the problem either, as far as I can tell.

My ideal would be to configure that the first health check is only done after the start period. That would allow me to use a start-period of 5 seconds, a check interval of 5 minutes and have the container go healthy within seconds.

Some sort of --wait-for-start option to trigger this behavior would be backward compatible and cover all described use cases, right?

rdxmb commented:
@cpuguy83 @dtakken  Ok, I think I got it.
I just tried to draft the suggestions made up to now: https://github.com/moby/moby/issues/33410#issuecomment-368694141

thaJeztah commented:
Discussing in the maintainers meeting: one option is to add a `--health-start-interval` instead, for example:

```
--health-start-interval=3sec
--health-interval=5min
--health-retries=3
--health-start-period=0
```

That configuration would:

- start the container
- start health-check every 3 seconds
- if three consecutive health-checks fail, the container fails

Combining this with start-period:


```
--health-start-interval=3sec
--health-interval=5min
--health-retries=3
--health-start-period=10min
```

This would:

- start the container
- perform health-checks every 3 seconds
- if the health-checks fail during the first 10 minutes (`--health-start-period`), health-checks continue
- if the health-check passes during the first 10 minutes, the container becomes healthy, and the interval is changing to run healthchecks every 5 minutes (`--health-interval`)


`--health-start-x` options by default use the same values as the regular `--health-x` options; In future we can add additional `--health-start-xx` options that would override those if we see a reason to add those.

ngrilly commented:
@thaJeztah The health-start-interval option is great! Is it still on the roadmap?

wolfi101 commented:
What's the current status here?
Without this fix a service will keep in "Starting" state until the first health check is performed.
This leads to a service downtime of the health intervale whenever a service is updated which makes the health checks unuseable if you don't want to have them in a high frequency.

cpuguy83 commented:
I've created #40894 to replace this.
